From 4d7b67d6e24a8121b4d9cc6b3976b01dede5b2be Mon Sep 17 00:00:00 2001
From: Leonid Rosenboim <lrosenboim@caviumnetworks.com>
Date: Sun, 12 May 2019 21:12:15 +0200
Subject: [PATCH 198/345] MIPS: Octeon: Get first 256MB from 32-bit addresable
 memory

This amount should satisfy the needs of SWIOTLB and drivers that
require 32-bit DMA buffers.

The need to explicitly specify the address range becomes obvious
when the cvmx_bootmem allocation policy is changed to "last fit",
preferring high memory to conserve low memory for clients that
really need it, and specify their address range explicitly.

Signed-off-by: Leonid Rosenboim <lrosenboim@caviumnetworks.com>
Signed-off-by: Carlos Munoz <cmunoz@caviumnetworks.com>
Signed-off-by: David Daney <david.daney@cavium.com>
---
 arch/mips/cavium-octeon/setup.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/arch/mips/cavium-octeon/setup.c b/arch/mips/cavium-octeon/setup.c
index 6cc5de661c32..3caebb1d9bd5 100644
--- a/arch/mips/cavium-octeon/setup.c
+++ b/arch/mips/cavium-octeon/setup.c
@@ -1085,8 +1085,15 @@ void __init plat_mem_setup(void)
 	while ((boot_mem_map.nr_map < BOOT_MEM_MAP_MAX)
 		&& (total < max_memory)) {
 #if defined(CONFIG_64BIT) || defined(CONFIG_64BIT_PHYS_ADDR)
+		const uint64_t min_low_ram = 256 << 20; /* 256 MBytes */
+		uint64_t limit;
+		if (total < min_low_ram)
+			limit = (1ull << 32)-1;
+		else
+			limit = ~0ull;
+
 		memory = cvmx_bootmem_phy_alloc(mem_alloc_size,
-						__pa_symbol(&_end), -1,
+						__pa_symbol(&_end), limit,
 						0x100000,
 						CVMX_BOOTMEM_FLAG_NO_LOCKING);
 #elif defined(CONFIG_HIGHMEM)
-- 
2.25.1

